src/http/Client.zig

```diff
@@ -680,88 +695,122 @@ pub const Transfer = struct {
 
         const header = buffer[0 .. buf_len - 2];
 
-        if (transfer.response_header == null) {
-            if (transfer._redirecting and buf_len == 2) {
-                // parse and set cookies for the redirection.
-                redirectionCookies(transfer, easy) catch |err| {
-                    log.debug(.http, "redirection cookies", .{ .err = err });
+        // transition the status dependending the previous one.
+        transfer._resp_header_status = switch (transfer._resp_header_status) {
+            .empty => .first,
+            .first => .next,
+            .next => .next,
+            .end => .first,
+        };
+
+        // mark the end of parsing headers
+        if (buf_len == 2) transfer._resp_header_status = .end;
+
+        switch (transfer._resp_header_status) {
+            .empty => unreachable,
+            .first => {
+                if (buf_len < 13) {
+                    log.debug(.http, "invalid response line", .{ .line = header });
+                    return 0;
+                }
+                const version_start: usize = if (header[5] == '2') 7 else 9;
+                const version_end = version_start + 3;
+
+                // a bit silly, but it makes sure that we don't change the length check
+                // above in a way that could break this.
+                std.debug.assert(version_end < 13);
+
+                const status = std.fmt.parseInt(u16, header[version_start..version_end], 10) catch {
+                    log.debug(.http, "invalid status code", .{ .line = header });
                     return 0;
                 };
-                return buf_len;
-            }
 
-            if (buf_len < 13 or std.mem.startsWith(u8, header, "HTTP/") == false) {
+                var url: [*c]u8 = undefined;
+                errorCheck(c.curl_easy_getinfo(easy, c.CURLINFO_EFFECTIVE_URL, &url)) catch |err| {
+                    log.err(.http, "failed to get URL", .{ .err = err });
+                    return 0;
+                };
+
+                // When using proxy, curl call the header function for all HTTP
+                // requests, including the CONNECT one used tunneling requests.
+                if (transfer._use_proxy and transfer.proxy_response_header == null) {
+                    transfer.proxy_response_header = .{
+                        .url = "",
+                        .status = status,
+                    };
+
+                    // We want to ignore the successful proxy's CONNECT request.
+                    // But there is no proper way to detect if the current
+                    // request is a proxy CONNECT one.
+                    // We know curl uses a CONNECT when it establishes a TLS
+                    // conn.
+                    if (status == 200 and std.mem.startsWith(u8, std.mem.span(url), "https")) {
```

Comment:
During response header callback, we don't have a reliable way to know the current request is a proxy `CONNECT` one.
`CURLINFO_HTTP_CONNECTCODE` would be the solution, but it is set only *after* the connect request performed.
That's why I rely on `https` prefix instead.